packages/swing-store/src/transcriptStore.js

```diff
@@ -400,6 +579,34 @@ export function makeTranscriptStore(
   `);
   sqlCountSpanItems.pluck();
 
+  const sqlGetSpansInRange = db.prepare(`
+    SELECT startPos, endPos
+    FROM transcriptSpans
+    WHERE vatID = ? AND startPos >= ? AND endPos <= ?
+    ORDER BY startPos
+  `);
+
+  function countSpanItems(vatID, rangeStartPos, rangeEndPos) {
+    // This is an expensive operation that should only be used for sanity checks
+    // during state sync import & export
+    let count = 0;
+    for (const { startPos, endPos } of sqlGetSpansInRange.iterate(
+      vatID,
+      rangeStartPos,
+      rangeEndPos,
+    )) {
+      const compressedSpanReader = readCompressedSpan(vatID, startPos);
+      if (compressedSpanReader) {
+        for (const _ of compressedSpanReader()) {
+          count += 1;
+        }
+      } else {
+        count += sqlCountSpanItems.get(vatID, startPos, endPos);
+      }
+    }
+    return count;
+  }
```

Comment:
I don't think we actually need to read the entities here, but since the prior version counted the entries as a sanity check, Brian and I looked at it decided we should do the same here.  Note that this is only happening in a case where the data is already being read and decompressed anyway.  If the cost here proves notable and we trust that our other code is working right (which, frankly, it kind of has to anyway), we could decide to dispense with this check.

The other thing is that we do end up doing extra compression & decompression (with our eyes open) because we'd like the compression-on-disk and the compression-over-the-wire for state sync to be decoupled.  If we changed that decision we could get rid of the decompression followed by compression on export/import and stream the compressed blobs directly.  But that would mean introducing dependency on the details of the compression scheme into the state sync protocol, which would probably be bad.

Reply:
> we'd like the compression-on-disk and the compression-over-the-wire for state sync to be decoupled. If we changed that decision we could get rid of the decompression followed by compression on export/import and stream the compressed blobs directly. But that would mean introducing dependency on the details of the compression scheme into the state sync protocol, which would probably be bad.

This is not entirely correct. While we don't want the state-sync artifacts to be compressed and dependent on a Node.js compression implementation (which may diverge between validators), the JS side is not responsible for this. We actually do want the export to be compressed throughout (see #8448), and instead would rely on the golang side to decompress the artifact it gets from JS during an export / snapshot creation. Internally the cosmos layer does recompress the state-sync data, but as a larger stream of protobuf messages, and we assume more deterministically.

Reply:
I tend to view clever SQL queries with an immense amount of suspicion.  But it's nice to switch to believing the metadata instead of decompressing and counting.

Reply:
Yeah, I'd be happy to have it be one DB query that gets all the relevant compressed span metadata (the list of spans), then a JS loop that queries `endPos-startPos` for each (one DB query per span), then a final DB query of `COUNT` for the loose items, and do all the addition in JS. It's a tradeoff between more-but-simpler DB queries vs fewer-but-more-clever DB queries. Maybe some day we should compare the perf of the two approaches, but this only happens during import, so I wouldn't investigate it until after we switch to compressed artifacts (which will be a bigger perf win, I think).